### PR TITLE
txpool: don't validate block transactions if the pool is empty

### PR DESCRIPTION
**Pull request by michalkucharczyk**
Fix shall prevent from wasting the CPU during the major sync. Block
transaction don't need to be re-validated when the txpool is empty.

Part of: #12903

### CODE REVIEW

**Review by lexnv — APPROVED** [review]
LGTM! 👍

**Comment by arkpar** [comment]
I'd prefer this check to be earlier in the event handling flow. This skips validation but there's still unnecessary work done to read potentially thousands of extrinsics from all blocks since the last notification into memory. 

I think the root of the issue is the way the tree route is queried here:
https://github.com/paritytech/substrate/blob/cad470b6b68d6d2c79371dd67d2f915b37ab8304/client/transaction-pool/src/lib.rs#L727

This code seems to calculate the route from a previous notification, expecting it to be short. In reality notifications are not guaranteed to be sent for all blocks.  I.e. if a node goes offline for a few days, then reconnects and catches up the route may be huge. The fact that we send a finality notification on each session end even when doing a full sync is not something the tx pool should rely on. There should be a reasonable maximum of blocks that are queried here.

**Comment by michalkucharczyk** [comment]
I agree this solution is not a perfect solution of original problem. Please refer to: https://github.com/paritytech/substrate/issues/12903#issuecomment-1357900859 for my other proposal.

> I.e. if a node goes offline for a few days, then reconnects and catches up the route may be huge.

- Just a thought - the local node being offline for long time, could have _non-propagate_ transactions in its txpool. It could build a local fork. To handle this case correctly and notify txs' listeners, the complete tree route must be computed, and all transaction from enacted/retracted blocks should be processed. Not sure if this can be avoided.
- question: is there any indication of this state in the client (sth like `is_major_syncing`). I see this: https://github.com/paritytech/substrate/blob/e118df551952d1aa7d85e79f04e1b8cb851b6761/client/network/common/src/sync.rs#L262, but not sure if this can be used as `SyncOracle`.

**Comment by arkpar** [comment]
> Just a thought - the local node being offline for long time, could have non-propagate transactions in its txpool. It could build a local fork. To handle this case correctly and notify txs' listeners, the complete tree route must be computed, and all transaction from enacted/retracted blocks should be processed. Not sure if this can be avoided.

The simplest solution I see is to simply clear the pool if the period between notifications is too large.

> question: is there any indication of this state (catching up after being offline) in the client (sth like is_major_syncing)

IIRC major syncing state indicates not just the initial sync, but also covers this case.

File: client/transaction-pool/src/graph/pool.rs [hunk 1]
```diff
@@ -271,14 +271,23 @@ impl<B: ChainApi> Pool<B> {
...
 				None => {
-					let validity = self
-						.validated_pool
-						.api()
-						.validate_transaction(parent, TransactionSource::InBlock, extrinsic.clone())
-						.await;
-
-					if let Ok(Ok(validity)) = validity {
-						future_tags.extend(validity.provides);
+					// Avoid validating block txs if the pool is empty (substrate/issues #12903)
```
Comment (bkchr) [inline client/transaction-pool/src/graph/pool.rs:274]:
```suggestion
					// Avoid validating block txs if the pool is empty
```

File: client/transaction-pool/src/graph/pool.rs [hunk 2]
```diff
@@ -271,14 +271,23 @@ impl<B: ChainApi> Pool<B> {
...
+								TransactionSource::InBlock,
+								extrinsic.clone(),
+							)
+							.await;
+
+						if let Ok(Ok(validity)) = validity {
+							future_tags.extend(validity.provides);
+						}
+					} else {
+						log::trace!( target: "txpool", "txpool is empty, skipping validation for block {:?}", at);
```
Comment (bkchr) [inline client/transaction-pool/src/graph/pool.rs:290]:
```suggestion
						log::trace!(target: "txpool", "txpool is empty, skipping validation for block {at:?}");
```

**Review by bkchr — APPROVED** [review]
While this doesn't solve the issue properly as @arkpar mentioned, this is still a good optimization.

**Comment by bkchr** [comment]
bot merge

**Pull request merged**